commcare_connect/opportunity/helpers.py

```diff
@@ -357,85 +357,148 @@ def get_payment_report_data(opportunity: Opportunity, usd=False):
-            default=Value(2),  # Inactive
-            output_field=IntegerField(),
-        ),
-    )
-    return queryset
-
-
-def get_opportunity_list_data(organization, program_manager=False):
-    today = now().date()
-    three_days_ago = now() - timedelta(days=3)
-
-    pending_approvals_sq = Subquery(
-        UserVisit.objects.filter(opportunity_access__opportunity_id=OuterRef("pk"), status="pending")
-        .values("opportunity_access__opportunity_id")
-        .annotate(count=Count("id", distinct=True))
-        .values("count")[:1],
-        output_field=IntegerField(),
-    )
-
-    base_filter = Q(organization=organization)
-    if program_manager:
-        base_filter |= Q(managedopportunity__program__organization=organization)
-
-    queryset = Opportunity.objects.filter(base_filter).annotate(
-        program=F("managedopportunity__program__name"),
-        pending_invites=pending_invites_subquery(),
-        pending_approvals=Coalesce(pending_approvals_sq, Value(0)),
-        total_accrued=total_accrued_sq(),
-        total_paid=total_paid_sq(),
-        payments_due=ExpressionWrapper(
-            F("total_accrued") - F("total_paid"),
-            output_field=DecimalField(),
-        ),
-        inactive_workers=inactive_workers_subquery(three_days_ago),
-        status=Case(
-            When(Q(active=True) & Q(end_date__gte=today), then=Value(0)),  # Active
-            When(Q(active=True) & Q(end_date__lt=today), then=Value(1)),  # Ended
-            default=Value(2),  # Inactive
-            output_field=IntegerField(),
-        ),
-    )
-
-    if program_manager:
-        total_deliveries_sq = Subquery(
-            CompletedWork.objects.filter(opportunity_access__opportunity_id=OuterRef("pk"))
-            .values("opportunity_access__opportunity_id")
-            .annotate(total=Sum("saved_completed_count"))
-            .values("total")[:1],
-            output_field=IntegerField(),
+class TieredQueryset:
+    # This queryset enables targetted count, order_by, pagination
+    #   for large querysets on which it may not be efficient to run
+    #   them.
+    #
+    # This works by passing a base_queryset that just returns ids
+    #   and data_qs_fn which returns the data for those ids
+    def __init__(self, base_qs, data_qs_fn):
+        self.base_qs = base_qs
+        self.data_qs_fn = data_qs_fn
+        self._ordered_qs = base_qs  # default unless order_by is called
+
+    def count(self):
+        return self.base_qs.count()
+
+    def __getitem__(self, key):
+        if isinstance(key, slice):
+            page_qs = self._ordered_qs[key]
+            ids = list(page_qs.values_list("id", flat=True))
+            return list(self.data_qs_fn(ids))
+        else:
+            obj = self._ordered_qs[key]
+            result = list(self.data_qs_fn([obj.id]))
+            if not result:
+                raise IndexError("Index out of range.")
```

Comment:
If the index was out of range, line 381 would raise that error right? This is a different error caused when an ID can't be found in the data queryset.

Reply:
Yeah, this could be removed.

This is handling the case where the base queryset is able to return a matching object ID, but the data queryset is not, which shouldn't really happen.